osgp/protocol-adapter-oslp/osgp-adapter-protocol-oslp-elster/src/main/java/org/opensmartgridplatform/adapter/protocol/oslp/elster/infra/messaging/DeviceRequestMessageProcessor.java

```diff
@@ -152,19 +154,32 @@ protected void handleError(final IOException e, final String correlationUid,
         this.responseMessageSender.send(protocolResponseMessage);
     }
 
+    protected void handleFunctionalException(final FunctionalException e, final MessageMetadata messageMetadata) {
+        LOGGER.error("Functional error while processing message, using MessageMetadata to create error response", e);
+
+        final ProtocolResponseMessage protocolResponseMessage = createProtocolResponseMessage(messageMetadata, e);
+        this.responseMessageSender.send(protocolResponseMessage);
+    }
+
     protected void handleError(final RuntimeException e, final MessageMetadata messageMetadata) {
         LOGGER.error("Error while processing message, using MessageMetadata to create error response", e);
 
         // Set the exception to a class known by all OSGP components
         final TechnicalException ex = new TechnicalException(ComponentType.PROTOCOL_OSLP, UNEXPECTED_EXCEPTION);
+        final ProtocolResponseMessage protocolResponseMessage = createProtocolResponseMessage(messageMetadata, ex);
+        this.responseMessageSender.send(protocolResponseMessage);
+    }
 
+    private ProtocolResponseMessage createProtocolResponseMessage(
+            final MessageMetadata messageMetadata, final OsgpException ex) {
         final DeviceMessageMetadata deviceMessageMetadata = new DeviceMessageMetadata(messageMetadata);
-        final ProtocolResponseMessage protocolResponseMessage = ProtocolResponseMessage.newBuilder()
-                .domain(messageMetadata.getDomain()).domainVersion(messageMetadata.getDomainVersion())
-                .deviceMessageMetadata(deviceMessageMetadata).result(ResponseMessageResultType.NOT_OK).osgpException(ex)
-                .retryCount(messageMetadata.getRetryCount()).build();
-
-        this.responseMessageSender.send(protocolResponseMessage);
+        return ProtocolResponseMessage.newBuilder()
+                    .domain(messageMetadata.getDomain())
+                    .domainVersion(messageMetadata.getDomainVersion())
+                    .deviceMessageMetadata(deviceMessageMetadata)
+                    .result(ResponseMessageResultType.NOT_OK)
+                    .osgpException(ex)
+                    .retryCount(messageMetadata.getRetryCount()).build();
     }
 
     public void handleUnableToConnectDeviceResponse(final DeviceResponse deviceResponse, final Throwable t,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Method has 7 parameters, which is greater than 6 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS00107)